File: scala/scala-impl/test/org/jetbrains/plugins/scala/lang/parser/stress/DragSearchTest.java
package org.jetbrains.plugins.scala.lang.parser.stress;

import com.intellij.lang.LanguageParserDefinitions;
import com.intellij.lang.ParserDefinition;
import com.intellij.lang.PsiBuilder;
import com.intellij.lang.PsiBuilderFactory;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Pair;
import com.intellij.openapi.util.TextRange;
import com.intellij.util.containers.ContainerUtil;
import junit.framework.Test;
import junit.framework.TestCase;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.plugins.scala.base.ScalaFileSetTestCase;
import org.junit.Assert;

public class DragSearchTest extends TestCase {
    private static final int MAX_ROLLBACKS = 30;

    private static void exploreForDrags(Pair<TextRange, Integer>[] dragInfo) {
        int ourMaximum = max(dragInfo);
        boolean notFound = ContainerUtil.findAll(dragInfo, pair -> pair.getSecond() >= MAX_ROLLBACKS).isEmpty();

        if (!notFound) {
            Assert.assertTrue("Too much rollbacks: " + ourMaximum, ourMaximum < MAX_ROLLBACKS);
        }

    }

    private static int max(Pair<TextRange, Integer>[] dragInfo) {
        int max = 0;
        for (Pair<TextRange, Integer> pair : dragInfo) {
            if (pair.getSecond() > max) {
                max = pair.getSecond();
            }
        }
        return max;
    }

    public static Test suite() {
        return new ScalaFileSetTestCase("/parser/stress/data/") {
            @Override
            protected void runTest(@NotNull String testName,
                                   @NotNull String content,
                                   @NotNull Project project) {
                transform(testName, content, project);
            }

            @Override
            @NotNull
            protected String transform(@NotNull String testName,
                                       @NotNull String fileText,
                                       @NotNull Project project) {
                ParserDefinition parserDefinition = LanguageParserDefinitions.INSTANCE.forLanguage(getLanguage());
                PsiBuilder psiBuilder = PsiBuilderFactory.getInstance().createBuilder(
                        parserDefinition,
                        parserDefinition.createLexer(project),
                        fileText
                );
                DragBuilderWrapper dragBuilder = new DragBuilderWrapper(project, psiBuilder);
                parserDefinition.createParser(project).parse(parserDefinition.getFileNodeType(), dragBuilder);

                Pair<TextRange, Integer>[] dragInfo = dragBuilder.getDragInfo();
                exploreForDrags(dragInfo);

                return super.transform(testName, fileText, project);
            }
        };
    }
}

File: scala/scala-impl/test/org/jetbrains/plugins/scala/lang/parser/stress/DragSearchTest.scala
package org.jetbrains.plugins.scala.lang.parser.stress

import com.intellij.lang.{LanguageParserDefinitions, PsiBuilderFactory}
import com.intellij.openapi.util.{Pair, TextRange}
import org.jetbrains.plugins.scala.base.{DefaultFileSetTestTransform, NoSdkFileSetTestBase}
import org.junit.Assert.assertTrue

import java.nio.file.Path

class DragSearchTest extends NoSdkFileSetTestBase with DefaultFileSetTestTransform {
  override protected def relativeTestDataPath: Path = Path.of("parser", "stress", "data")

  override protected def runTest(testName: String, fileText: String): Unit = {
    transform(testName, fileText)
  }

  override protected def transform(testName: String, fileText: String): String = {
    val parserDefinition = LanguageParserDefinitions.INSTANCE.forLanguage(language)
    val psiBuilder = PsiBuilderFactory.getInstance().createBuilder(
      parserDefinition,
      parserDefinition.createLexer(project),
      fileText
    )
    val dragBuilder = new DragBuilderWrapper(project, psiBuilder)
    parserDefinition.createParser(project).parse(parserDefinition.getFileNodeType(), dragBuilder)

    val dragInfo = dragBuilder.getDragInfo.map(p2p)
    exploreForDrags(dragInfo)

    super.transform(testName, fileText)
  }

  private final val MaxRollbacks = 30

  private def exploreForDrags(dragInfo: Array[(TextRange, Int)]): Unit = {
    val ourMaximum = dragInfo.maxBy(_._2)._2
    val notFound = !dragInfo.exists(_._2 >= MaxRollbacks)

    if (!notFound) {
      assertTrue("Too many rollbacks: " + ourMaximum, ourMaximum < MaxRollbacks)
    }
  }

  private def p2p(pair: Pair[TextRange, Integer]): (TextRange, Int) = (pair.getFirst, pair.getSecond)
}
